native/swift/Sources/wordpress-api-cache/DatabaseChangeNotifier.swift

```diff
@@ -0,0 +1,294 @@
+    ///   - key: The entity ID to associate with the token.
+    ///   - replacementCallback: A closure called with the existing token if one exists.
+    func add(_ token: NSObjectProtocol, for key: EntityId, replacementCallback: (NSObjectProtocol) -> Void) {
+        entityLock.withLock {
+            if let existingToken = self.entityTokens[key] {
+                replacementCallback(existingToken)
+            }
+
+            self.entityTokens[key] = token
+        }
+    }
+
+    /// Adds an observer token for a specific collection.
+    ///
+    /// If a token already exists for the given collection handle, the replacement callback is called
+    /// with the existing token (typically to remove the observer), then the new token replaces it.
+    ///
+    /// - Parameters:
+    ///   - token: The notification observer token to store.
+    ///   - key: The collection handle to associate with the token.
+    ///   - replacementCallback: A closure called with the existing token if one exists.
+    func add(_ token: NSObjectProtocol, for key: UInt64, replacementCallback: (NSObjectProtocol) -> Void) {
+        collectionLock.withLock {
+            if let existingToken = self.collectionTokens[key] {
+                replacementCallback(existingToken)
+            }
+
+            self.collectionTokens[key] = token
+        }
+    }
+
+    /// Retrieves and removes an observer token for a specific entity.
+    ///
+    /// This method atomically retrieves the token and removes it from storage.
+    /// If no token exists for the given entity ID, returns `nil`.
+    ///
+    /// - Parameter key: The entity ID to look up.
+    /// - Returns: The observer token if one exists, or `nil` otherwise.
+    func get(for key: EntityId) -> NSObjectProtocol? {
+        entityLock.withLock {
+            guard let value = self.entityTokens[key] else {
+                return nil
+            }
+
+            self.entityTokens.removeValue(forKey: key)
+            return value
+        }
+    }
+
+    /// Retrieves and removes an observer token for a specific collection.
+    ///
+    /// This method atomically retrieves the token and removes it from storage.
+    /// If no token exists for the given collection handle, returns `nil`.
+    ///
+    /// - Parameter key: The collection handle to look up.
+    /// - Returns: The observer token if one exists, or `nil` otherwise.
+    func get(for key: UInt64) -> NSObjectProtocol? {
+        collectionLock.withLock {
+            guard let value = self.collectionTokens[key] else {
+                return nil
+            }
+
+            self.collectionTokens.removeValue(forKey: key)
+            return value
+        }
+    }
+}
+
+public struct UpdateHookSequence: AsyncSequence, Sendable {
+
+    let name: Notification.Name
+
+    public struct AsyncIterator : AsyncIteratorProtocol {
+        var publisher: AsyncCompactMapSequence<AsyncPublisher<NotificationCenter.Publisher>, UpdateHook>.Iterator
```

Comment:
I don't think you need `UpdateHookSequence`, it's just a wrapper around this async publisher?

If you want to erase the type, you can use `any AsyncSequence<UpdateHook, Never>` as the return value.

If [the `Failure` associated type is introduced on iOS 18 and macOS 15](https://developer.apple.com/documentation/swift/asyncsequence/failure) causes some compilation issues, maybe we simply return `AnyPublisher<UpdateHook, Never>`, and let the client call the `.values`?